File: year-2022/day-1/solve.py
import sys
import re

def main(args):
    pre = lambda x: int(x, 10)
    if "-ex" in args:
        elves = getLines("data.ex.txt", pred = pre)
    else:
        elves = getLines("data.in.txt", pred = pre)
    print(f"part 1: {elves[-1]}")
    print(f"part 2: {sum(elves[-3:])}")

def getSums(l):
  pass

def getLines(fn, **kw):
    pred = (lambda x: x) if "pred" not in kw else kw["pred"]
    output = []
    with open(fn) as f:
        line = f.readline()
        output.append(0)
        while line:
            trimLine = line[:-1]
            if trimLine:
              output[-1] += int(trimLine)
            else:
              output.append(0)
            line = f.readline()
    return sorted(output)

if __name__ == '__main__':
    main(sys.argv)

File: year-2022/day-1/solve.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

#include <algorithm>
#include <map>
#include <vector>

#define IO_USE                                                                 \
  using std::cout;                                                             \
  using std::cin;                                                              \
  using std::endl;                                                             \
  using std::string;

using argmap = std::map<std::string, std::string>;

argmap getArgs(int argc, char const *argv[]) {
  if (argc == 1) {
    return argmap();
  }
  argmap comms;
  bool parsing = false, first = true;
  std::string k, v;
  for (int i = 1; i < argc; ++i) {
    if (argv[i][0] == '-') {
      if (!first) {
        comms.emplace(k, v);
      }
      first = false;
      k = std::string(argv[i]);
      v = "";
      parsing = true;
    } else if (parsing) {
      v += std::string(argv[i]);
    }
  }
  comms.emplace(k, v);
  return comms;
}

bool hasKey(std::string key, argmap args) {
  return args.find(key) != args.end();
}

int main(int argc, char const *argv[]) {
  IO_USE;
  const argmap params = getArgs(argc, argv);
  std::string filename = "data.in.txt";
  if (hasKey("-f", params)) {
    filename = params.at("-f");
  };
  std::ifstream infile(filename);
  string line;
  std::vector<int> elves;
  elves.push_back(0);
  std::vector<int> largest;

  while (std::getline(infile, line)) {
    std::istringstream iss(line);
    int cal = 0;
    if (line.size()) {
      iss >> cal;
      elves.back() += cal;
    } else {
      elves.push_back(0);
    }
  }

  int top3 = 0;
  std::sort(elves.begin(), elves.end());
  cout << "part 1: " << largest.back() << endl;
  for (size_t i = 0; i < 3; ++i) {
    cout << elves.back() << endl;
    top3 += elves.back();
    elves.pop_back();
  }
  cout << "part 2: " << top3 << endl;
  return 0;
}
